janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -434,43 +444,43 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
     private class CQLPagingIterator implements Iterator<Row> {
 
         private ResultSet currentResultSet;
+        private java.util.Iterator<Row> currentPageIterator;
 
-        private int index;
-        private int paginatedResultSize;
-        private final Supplier<Statement> statementSupplier;
+        private final Supplier<BatchableStatement<BoundStatement>> statementSupplier;
 
-        private byte[] lastPagingState = null;
-
-        public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSupplier) {
-            this.index = 0;
-            this.paginatedResultSize = pageSize;
+        public CQLPagingIterator(Supplier<BatchableStatement<BoundStatement>> statementSupplier) {
             this.statementSupplier = statementSupplier;
-            this.currentResultSet = getResultSet();
+            updateResultSet();
         }
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            if(currentPageIterator.hasNext()){
+                return true;
+            }
+            if(currentResultSet.isFullyFetched()){
+                return false;
+            }
+            updateResultSet();
```

Comment:
Do you know when this line will be executed? Codecov shows this is not covered. By testing locally, I created a graph with half a million vertices and did a full scan, but still, this line was not executed.

Reply:
Hmm. Interesting. This line should be executed when `currentPageIterator` has no any elements on this page but there are still some pages available i.e. `!currentPageIterator.hasNext() && !currentResultSet.isFullyFetched()`

Reply:
Did you have `1,000,000` vertices in `currentPageIterator`?

Reply:
What was your `storeManager.getPageSize()`?

Reply:
Interesting. This line isn't reachable for me as well.

Reply:
I will check that again. Looks like pages are fetched automatically but as far as I remember it wasn't true for the new driver. Maybe something changed or I missed something and pages are fetched automatically in the iterator. I will check that. Nevertheless this logic shouldn't break anything but if this logic is redundant we should remove it to simplify it.

Reply:
Okay, I checked, indeed DSE driver automatically fetches pages but now I see why this iterator was implemented. It has a comment above that automatic DSE pagination could lead to OOM because it was storing the whole history. Thus, we were adding another level of pagination. I will check if it is the issue still and if so will fix this part of the logic. Thank you for catching this!

Reply:
`ExecutionInfo` is stored for each new page as seen here:
https://github.com/datastax/java-driver/blob/c225588811a85e25d98790ce4b4f399e0ca066e4/core/src/main/java/com/datastax/dse/driver/internal/core/graph/MultiPageGraphResultSet.java#L108

So, potentially it may result in OOM. Thus, I'm changing the logic of this iterator back to forcing new request per page as it was before. It's great that you noticed this issue @li-boxuan ! Thanks!

Reply:
Fixed with explanations here: https://github.com/JanusGraph/janusgraph/pull/2169#issuecomment-730041434